src/credentials/FabricTable.cpp

```diff
@@ -41,145 +42,156 @@ CHIP_ERROR FabricInfo::SetFabricLabel(const CharSpan & fabricLabel)
+constexpr TLV::Tag kFabricLabelTag = TLV::ContextTag(1);
+} // anonymous namespace
+
 CHIP_ERROR FabricInfo::CommitToStorage(PersistentStorageDelegate * storage)
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
+    DefaultStorageKeyAllocator keyAlloc;
 
-    char key[kKeySize];
-    ReturnErrorOnFailure(GenerateKey(mFabric, key, sizeof(key)));
+    VerifyOrReturnError(CanCastTo<uint16_t>(mRootCert.size()) && CanCastTo<uint16_t>(mICACert.size()) &&
+                            CanCastTo<uint16_t>(mNOCCert.size()),
+                        CHIP_ERROR_BUFFER_TOO_SMALL);
 
-    StorableFabricInfo * info = chip::Platform::New<StorableFabricInfo>();
-    ReturnErrorCodeIf(info == nullptr, CHIP_ERROR_NO_MEMORY);
+    ReturnErrorOnFailure(
+        storage->SyncSetKeyValue(keyAlloc.FabricRCAC(mFabric), mRootCert.data(), static_cast<uint16_t>(mRootCert.size())));
 
-    info->mFabricIndex = mFabric;
-    info->mVendorId    = Encoding::LittleEndian::HostSwap16(mVendorId);
+    // If we stop storing ICA certs when empty, update LoadFromStorage
+    // accordingly to check for CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND.
+    ReturnErrorOnFailure(
+        storage->SyncSetKeyValue(keyAlloc.FabricICAC(mFabric), mICACert.data(), static_cast<uint16_t>(mICACert.size())));
 
-    size_t stringLength = strnlen(mFabricLabel, kFabricLabelMaxLengthInBytes);
-    memcpy(info->mFabricLabel, mFabricLabel, stringLength);
-    info->mFabricLabel[stringLength] = '\0'; // Set null terminator
+    ReturnErrorOnFailure(
+        storage->SyncSetKeyValue(keyAlloc.FabricNOC(mFabric), mNOCCert.data(), static_cast<uint16_t>(mNOCCert.size())));
 
+    Crypto::P256SerializedKeypair serializedOpKey;
     if (mOperationalKey != nullptr)
     {
-        SuccessOrExit(err = mOperationalKey->Serialize(info->mOperationalKey));
+        ReturnErrorOnFailure(mOperationalKey->Serialize(serializedOpKey));
     }
     else
     {
+        // Could we just not store it instead?  What would deserialize need to
+        // do then?
         P256Keypair keypair;
-        SuccessOrExit(err = keypair.Initialize());
-        SuccessOrExit(err = keypair.Serialize(info->mOperationalKey));
+        ReturnErrorOnFailure(keypair.Initialize());
+        ReturnErrorOnFailure(keypair.Serialize(serializedOpKey));
     }
 
-    if (mRootCert.empty())
-    {
-        info->mRootCertLen = 0;
-    }
-    else
-    {
-        VerifyOrExit(CanCastTo<uint16_t>(mRootCert.size()), err = CHIP_ERROR_INVALID_ARGUMENT);
-        info->mRootCertLen = Encoding::LittleEndian::HostSwap16(static_cast<uint16_t>(mRootCert.size()));
-        memcpy(info->mRootCert, mRootCert.data(), mRootCert.size());
-    }
+    const auto opKeyLength = serializedOpKey.Length();
+    VerifyOrReturnError(CanCastTo<uint16_t>(opKeyLength), CHIP_ERROR_BUFFER_TOO_SMALL);
+    ReturnErrorOnFailure(
+        storage->SyncSetKeyValue(keyAlloc.FabricOpKey(mFabric), serializedOpKey.Bytes(), static_cast<uint16_t>(opKeyLength)));
```

Comment:
We should probably version this serialized key with TLV, since it's quite an opaque format. We do not want to be moving forward in core SDK code in a way that a crypto backend changes the internal representation, and then it blows-up later when trying to load, but you can't tell why...

Reply:
Let me think about the options here.

Reply:
Did a TLV bit with a tag for version and tag for data.